README.md

# bio-curve-fit

A Python package for fitting common dose-response and standard curve models. Designed to follow the [scikit-learn](https://scikit-learn.org/stable/) api.

## Quickstart 

### Installation

```shell
pip install bio-curve-fit
```

### Example usage:

```python
from bio_curve_fit.logistic import FourPLLogistic

# Instantiate model
model = FourPLLogistic()

# create some example data
standard_concentrations = [1, 2, 3, 4, 5]
standard_responses = [0.5, 0.55, 0.9, 1.25, 1.55]


# fit the model using an optional inverse variance weight function (1/y^2)
model = FourPLLogistic().fit(
	standard_concentrations, 
	standard_responses, 
)

# interpolate the response at given concentrations
model.predict([1.5, 2.5])

# interpolate the concentration at given responses
model.predict_inverse([0.1, 1.0])

```

Calculate and plot the curve and limits of detection:

```python
plot_standard_curve(standard_concentrations, standard_responses, model, show_plot=True)
```

![standard curve](./examples/readme_fit.png)

## Examples

See the [example notebook](./examples/four_pl_logistic/four_pl_fit.ipynb) for more detailed usage.

## Contributing

Contributions are welcome! We built this package to be useful for our own work, but we know there is more to add.
Please see [CONTRIBUTING.md](CONTRIBUTING.md) for more information.


CHANGELOG.md
# Changelog
All notable changes to this packages will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Prediction intervals method
### Fixed
- Jacobian calculation for 4PL model
### Changed
- Modified 4PL formula to avoid warning associated with complex numbers
- Renamed std dev method to confidence interval to clarify meaning
- Split fit based on whether model is positively or negatively sloping
### Deprecated
### Removed
### Security

tests/test_four_pl_logistic.py
import numpy as np
import pandas as pd
import pytest

from bio_curve_fit.logistic import FourPLLogistic
from bio_curve_fit.plotting import plot_standard_curve

# set a seed for reproducibility
np.random.seed(42)


def test_fit_and_plot():
    TEST_PARAMS = [1.0, 1.0, 2.0, 3.0]

    x_data = np.logspace(0.00001, 7, 100, base=np.e)  # type: ignore
    # generate y-data based on the test parameters
    y_data = FourPLLogistic.four_param_logistic(
        x_data + np.random.normal(0.0, 0.1 * x_data, len(x_data)), *TEST_PARAMS
    )

    model = FourPLLogistic().fit(
        x_data, y_data, weight_func=FourPLLogistic.inverse_variance_weight_function
    )

    # model should recover parameters used to generate the data
    params = list(model.get_params().values())
    assert np.isclose(params, TEST_PARAMS, rtol=0.4).all()  # type: ignore

    r2 = model.score(x_data, y_data)
    assert r2 > 0.995

    # test plotting
    plot_standard_curve(x_data, y_data, model)

    # test __repr__
    print(model)


test_y = pd.Series(
    [
        223747,
        214105,
        61193,
        61831,
        16290,
        14151,
        4097,
        3895,
        1587,
        1485,
        896,
        771,
        743,
        674,
        653,
        562,
        634,
        642,
    ]
)

test_x = pd.Series(
    [
        1360,
        1360,
        340,
        340,
        85,
        85,
        21.25,
        21.25,
        5.3125,
        5.3125,
        1.328125,
        1.328125,
        0.33203125,
        0.33203125,
        0.0830078125,
        0.0830078125,
        0.0,
        0.0,
    ]
)


def test_fit2():
    model = FourPLLogistic().fit(
        test_x,
        test_y,
        weight_func=FourPLLogistic.inverse_variance_weight_function,
    )
    print("Params:", model.get_params())
    print(model.predict_inverse(0.1))
    plot_standard_curve(test_x, test_y, model)
    assert model.score(test_x, test_y) > 0.995  # type: ignore
    print(model.ULOD_y_, model.LLOD_y_)

    assert np.isclose(model.ULOD_y_, 220006.8397685415, rtol=0.1)  # type: ignore
    assert np.isclose(model.LLOD_y_, 798.7000577483678, rtol=0.1)  # type: ignore


@pytest.mark.filterwarnings("ignore::UserWarning")
def test_readme_example():
    """
    ensure that the example in the README works
    """
    # Instantiate model
    model = FourPLLogistic()

    # create some example data
    standard_concentrations = [1, 2, 3, 4, 5]
    standard_responses = [0.5, 0.55, 0.9, 1.25, 1.55]

    # fit the model using an inverse variance weight function (1/y^2)
    model = FourPLLogistic().fit(
        standard_concentrations,
        standard_responses,
    )

    # interpolate the response at given concentrations
    values = model.predict([1.5, 2.5])
    assert pd.notna(values).all()

    # interpolate the concentration at given responses
    values = model.predict_inverse([0.1, 1.0])
    assert pd.notna(values).all()
    plot_standard_curve(standard_concentrations, standard_responses, model)


def test_std_dev():
    model = FourPLLogistic().fit(
        test_x,
        test_y,
    )
    ci = model.predict_confidence_band(test_x)
    pi = model.predict_prediction_band(test_x, test_y)

    print(ci)
    print(pi)


bio_curve_fit/base.py
from abc import ABC, abstractmethod
from typing import Optional


class BaseStandardCurve(ABC):
    """
    Interface for standard curve models.
    """

    # Upper and lower Limits of Detection ("LODs")
    # Estimated Limits of Detection for concentration
    LLOD: Optional[float]
    ULOD: Optional[float]
    # Estimated Limits of Detection for response signal
    LLOD_y_: Optional[float]
    ULOD_y_: Optional[float]

    @abstractmethod
    def predict(self, x):
        return None

    @abstractmethod
    def predict_inverse(self, y):
        """
        Predict the inverse (x-value for a given y-value).

        Often standard curves are fit to data where the x-values are known
        concentrations an y-values are measured responses. Later, we usually wish
        to predict the concentration for a given response (i.e. x given y)
        """
        pass


bio_curve_fit/plotting.py
import io

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from matplotlib.ticker import ScalarFormatter

from .base import BaseStandardCurve


def plot_standard_curve(
    x_data,
    y_data,
    fitted_model: BaseStandardCurve,
    title="4PL Curve Fit",
    x_label="Concentration",
    y_label="Response",
    show_plot: bool = False,
) -> bytes:
    """
    Generate a plot of the data and the fitted curve.
    """
    # Plot the data and the fitted curve
    # set x-axis to log scale
    # set scales to log
    plt.xscale("log")
    plt.yscale("log")
    data = pd.DataFrame({"x": x_data, "y": y_data})
    # remove zeros from x_data
    filtered_data = data[data["x"] > 0]

    # Plot the fitted curve
    epsilon = 0.01
    x_min = np.log10(max(min(x_data), epsilon))
    x_max = max(x_data) * 2
    x = np.logspace(x_min, np.log10(x_max), 100)  # type: ignore
    # Generate y-data based on the fitted parameters
    y_pred = fitted_model.predict(x)

    plt.plot(x, y_pred, label="Fitted curve", color="red")
    plt.scatter(filtered_data["x"], filtered_data["y"], label="Data", s=12)
    formatter = ScalarFormatter()
    formatter.set_scientific(False)
    plt.gca().xaxis.set_major_formatter(formatter)
    plt.xlabel(x_label)
    plt.ylabel(y_label)
    plt.title(title)

    # set horizontal and vertical lines for ULOD and LLOD
    llod_response, ulod_response = fitted_model.LLOD_y_, fitted_model.ULOD_y_
    plt.axhline(llod_response, color="red", linestyle="--", label="LLOD")  # type: ignore
    plt.axhline(ulod_response, color="blue", linestyle="--", label="ULOD")  # type: ignore
    plt.legend()
    if show_plot:
        plt.show()
    # Save the plot to a BytesIO object
    buf = io.BytesIO()
    plt.savefig(buf, format="png")
    plt.clf()
    buf.seek(0)
    return buf.read()


bio_curve_fit/four_pl_logistic.py
from typing import Optional

import numpy as np
import pandas as pd
from scipy.optimize import curve_fit  # type: ignore
from sklearn.base import BaseEstimator, RegressorMixin  # type: ignore


class FourPLLogistic(BaseEstimator, RegressorMixin):
    def __init__(
        self,
        A=None,
        B=None,
        C=None,
        D=None,
        LLOD=None,
        ULOD=None,
        ULOD_y=None,
        LLOD_y=None,
        slope_direction_positive: Optional[bool] = None,
        slope_guess_num_points_to_use: int = 3,
    ):
        # A is the minimum asymptote
        self.A_ = A
        # B is the Hill's slope
        self.B_ = B
        # C is the inflection point (EC50)
        self.C_ = C
        # D is the maximum asymptote
        self.D_ = D
        self.cov_ = None
        # Initial guesses used when fitting the curve
        self.guess_A_ = None
        self.guess_B_ = None
        self.guess_C_ = None
        self.guess_D_ = None
        # Estimated Limits of Detection for response signal
        self.LLOD_y_ = LLOD_y
        self.ULOD_y_ = ULOD_y
        # Estimated Limits of Detection for concentration
        self.LLOD_ = LLOD
        self.ULOD_ = ULOD
        self.slope_direction_positive = slope_direction_positive
        self.slope_guess_num_points_to_use = slope_guess_num_points_to_use

    def get_params(self, deep=False):
        if deep:
            return {
                "A": self.A_,
                "B": self.B_,
                "C": self.C_,
                "D": self.D_,
                "LLOD": self.LLOD_,
                "ULOD": self.ULOD_,
                "ULOD_y": self.ULOD_y_,
                "LLOD_y": self.LLOD_y_,
            }
        else:
            return {
                "A": self.A_,
                "B": self.B_,
                "C": self.C_,
                "D": self.D_,
            }

    @staticmethod
    def four_param_logistic(x, A, B, C, D):
        """4 Parameter Logistic (4PL) model."""

        # For addressing fractional powers of negative numbers
        # https://stackoverflow.com/questions/45384602/numpy-runtimewarning-invalid-value-encountered-in-power
        z = np.sign(x / C) * np.abs(x / C) ** B

        return ((A - D) / (1.0 + z)) + D

    @staticmethod
    def inverse_variance_weight_function(y_data):
        """
        Function for weighting residuals by 1/y^2 in `scipy.optimize.curve_fit`.
        """
        # To avoid division by zero, add a small constant to y_data.
        return y_data + np.finfo(float).eps

    def _calculate_lod_replicate_variance(
        self,
        x_data,
        y_data,
        lower_std_dev_multiplier: float = 2.5,
        upper_std_dev_multiplier: float = 0.0,
    ):
        """
        Calculate the Lower and Upper Limits of Detection (LLOD and ULOD) using variance
        of replicate max and min concentration standards. It ignore zero concentration
        standards. If there are no replicates, the standard deviation zero
        TODO: some minimum variance should be used.

        In the notation below we assume the response signal is the Y-axis and the
        concentration is the X-axis.

        Example: Two replicates of the lowest concentration standard (conc=1.0 pg/ml)
        have standard deviation of 100 across their responses. LLOD will be `model.predict
        (1.0) + 100 * 2.5` where 2.5 is the `lower_std_dev_multiplier` parameter.

        :param bottom_std_dev: Standard deviation at the bottom calibration point.
        :param top_std_dev: Standard deviation at the top calibration point.
        :param std_dev_multiplier: Multiplier for the standard deviations (default 2.5).
        :return: Pair of tuples containing the LLOD and ULOD, and the corresponding x-values.
        """

        x_indexed_y_data = pd.DataFrame({"x": x_data, "y": y_data}).set_index("x")
        # remove zeros from x_data
        x_indexed_y_data = x_indexed_y_data[x_indexed_y_data.index > 0]  # type: ignore
        x_min = np.min(x_indexed_y_data.index)  # type: ignore
        x_max = np.max(x_indexed_y_data.index)  # type: ignore
        bottom_std_dev = x_indexed_y_data.loc[x_min, "y"].std()  # type: ignore
        top_std_dev = x_indexed_y_data.loc[x_max, "y"].std()  # type: ignore

        # Calculate LLOD and ULOD of RESPONSE SIGNAL
        llod = self.predict(x_min) + (lower_std_dev_multiplier * bottom_std_dev)
        ulod = self.predict(x_max) - (upper_std_dev_multiplier * top_std_dev)

        # Calculate the limits of detection for CONCENTRATION
        llod_x = self.predict_inverse(llod)
        ulod_x = self.predict_inverse(ulod)
        return llod_x, ulod_x, llod, ulod

    def fit(self, x_data, y_data, weight_func=None, LOD_func=None, **kwargs):
        """
        Fit the 4 Parameter Logistic (4PL) model.

        x_data: x data points
        y_data: y data points
        weight_func: optional Function that calculates weights from y_data. This is
            passed into the `curve_fit` function where the function minimized is `sum
            ((r / weight_func(y_data)) ** 2)` where r is the residuals.
            Thus for a typical 1/y^2 weighting, `weight_func` should be `lambda
            y_data: y_data`
        """
        x_data = np.float64(x_data)
        y_data = np.float64(y_data)
        df_data = pd.DataFrame({"x": x_data, "y": y_data})
        df_data.sort_values(by="x", inplace=True)

        if LOD_func is None:
            # default LOD_func is to use replicate variance
            LOD_func = self._calculate_lod_replicate_variance

        absolute_sigma = False
        weights = None
        if weight_func is not None:
            weights = weight_func(y_data)
            absolute_sigma = True

        # Initial guess for the parameters
        self.guess_A_ = np.min(y_data)
        if self.slope_direction_positive is not None:
            self.guess_B_ = 1.0 if self.slope_direction_positive else -1.0
        else:
            self.guess_B_ = (
                1.0
                if np.mean(
                    df_data.iloc[: np.minimum(self.slope_guess_num_points_to_use, len(df_data))][
                        "y"
                    ]
                )
                < np.mean(
                    df_data.iloc[-np.minimum(self.slope_guess_num_points_to_use, len(df_data)) :][
                        "y"
                    ]
                )
                else -1.0
            )
        self.guess_C_ = np.mean(x_data)
        self.guess_D_ = np.max(y_data)
        initial_guess = [self.guess_A_, self.guess_B_, self.guess_C_, self.guess_D_]

        curve_fit_kwargs = {
            "f": self.four_param_logistic,
            "xdata": x_data,
            "ydata": y_data,
            "p0": initial_guess,
            "maxfev": 10000,
            # jac=self.jacobian,
            "sigma": weights,
            "absolute_sigma": absolute_sigma,
        }

        # overwrite parameters with any kwargs passed in
        for k, v in kwargs.items():
            curve_fit_kwargs[k] = v

        # Perform the curve fit
        params, cov = curve_fit(**curve_fit_kwargs)
        self.A_, self.B_, self.C_, self.D_ = params
        self.cov_ = cov
        self.LLOD_, self.ULOD_, self.LLOD_y_, self.ULOD_y_ = LOD_func(x_data, y_data)
        return self

    @staticmethod
    def jacobian(x_data, A, B, C, D):
        """
        Jacobian matrix of the 4PL function with respect to A, B, C, D.
        """
        z = (x_data / C) ** B

        partial_A = 1.0 / (1.0 + z)
        partial_B = -(z * (A - D) * np.log(np.maximum(x_data / C, np.finfo(float).eps))) / (
            (1.0 + z) ** 2
        )
        partial_C = (B * z * (A - D)) / (C * (1.0 + z) ** 2)
        partial_D = 1.0 - 1.0 / (1.0 + z)

        # Jacobian matrix
        J = np.array([partial_A, partial_B, partial_C, partial_D]).T
        return J

    def predict_confidence_band(self, x_data):
        """
        Predict confidence bands of data points.
        TODO: still need to double-check the math here.

        See:
            https://www.graphpad.com/guides/prism/latest/curve-fitting/reg_graphing_confidence_and_predic.htm
            https://www.graphpad.com/guides/prism/latest/curve-fitting/reg_how_confidence_and_prediction_.htm
            https://stats.stackexchange.com/questions/15423/how-to-compute-prediction-bands-for-non-linear-regression

        """
        if self.cov_ is None:
            raise Exception(
                "Covariance matrix is not available. Please call 'fit' with appropriate data."
            )
        J = self.jacobian(x_data, self.A_, self.B_, self.C_, self.D_)
        pred_var = np.sum((J @ self.cov_) * J, axis=1)

        return np.sqrt(pred_var)

    def predict_prediction_band(self, x_data, y_data):
        """
        Predict prediction bands of data points.
        TODO: still need to double-check the math here.
        """
        ss = (y_data - self.predict(x_data)) ** 2
        df = len(x_data) - 4  # 4 parameters

        return np.sqrt(self.predict_confidence_band(x_data) ** 2 * ss / df)

    def predict_inverse(self, y):
        # TODO: could merge this with `predict` and make it a parameter
        """Inverse 4 Parameter Logistic (4PL) model.

        Used for calculating the x-value for a given y-value.
        Usually, standard curves are fitted using concentration as x-values and response as
        y-values, so that variance in response is modeled for a given known concentration.
        But for samples of unknown concentration, we want to get the concentration as given
        response, which is what this function does.

        """
        z = ((self.A_ - self.D_) / (y - self.D_)) - 1

        # For addressing fractional powers of negative numbers, np.sign(z) * np.abs(z) used rather than z
        # https://stackoverflow.com/questions/45384602/numpy-runtimewarning-invalid-value-encountered-in-power
        return self.C_ * (np.sign(z) * np.abs(z) ** (1 / self.B_))  # type: ignore

    def predict(self, x_data):
        if self.A_ is None or self.B_ is None or self.C_ is None or self.D_ is None:
            raise Exception("Model is not fitted yet. Please call 'fit' with appropriate data.")

        return self.four_param_logistic(x_data, self.A_, self.B_, self.C_, self.D_)


bio_curve_fit/logistic.py
from typing import Optional

import numpy as np
import pandas as pd
from scipy.optimize import curve_fit
from sklearn.base import BaseEstimator, RegressorMixin

from .base import BaseStandardCurve


class FourPLLogistic(BaseEstimator, RegressorMixin, BaseStandardCurve):
    def __init__(
        self,
        A=None,
        B=None,
        C=None,
        D=None,
        LLOD=None,
        ULOD=None,
        ULOD_y=None,
        LLOD_y=None,
        slope_direction_positive: Optional[bool] = None,
        slope_guess_num_points_to_use: int = 3,
    ):
        # A is the minimum asymptote
        self.A_ = A
        # B is the Hill's slope
        self.B_ = B
        # C is the inflection point (EC50)
        self.C_ = C
        # D is the maximum asymptote
        self.D_ = D
        self.cov_ = None
        # Initial guesses used when fitting the curve
        self.guess_A_ = None
        self.guess_B_ = None
        self.guess_C_ = None
        self.guess_D_ = None
        # Estimated Limits of Detection for response signal
        self.LLOD_y_ = LLOD_y
        self.ULOD_y_ = ULOD_y
        # Estimated Limits of Detection for concentration
        self.LLOD_ = LLOD
        self.ULOD_ = ULOD
        self.slope_direction_positive = slope_direction_positive
        self.slope_guess_num_points_to_use = slope_guess_num_points_to_use

    def check_fit(self):
        if self.A_ is None or self.B_ is None or self.C_ is None or self.D_ is None:
            raise Exception(
                "Model is not fit yet. Please call 'fit' with appropriate data"
                " or initialize the model object with non-null parameters."
            )

    def get_params(self, deep=False):
        if deep:
            return {
                "A": self.A_,
                "B": self.B_,
                "C": self.C_,
                "D": self.D_,
                "LLOD": self.LLOD_,
                "ULOD": self.ULOD_,
                "ULOD_y": self.ULOD_y_,
                "LLOD_y": self.LLOD_y_,
            }
        else:
            return {
                "A": self.A_,
                "B": self.B_,
                "C": self.C_,
                "D": self.D_,
            }

    @staticmethod
    def four_param_logistic(x, A, B, C, D):
        """4 Parameter Logistic (4PL) model."""

        # For addressing fractional powers of negative numbers
        # https://stackoverflow.com/questions/45384602/numpy-runtimewarning-invalid-value-encountered-in-power
        z = np.sign(x / C) * np.abs(x / C) ** B

        return ((A - D) / (1.0 + z)) + D

    @staticmethod
    def inverse_variance_weight_function(y_data):
        """
        Function for weighting residuals by 1/y^2 in `scipy.optimize.curve_fit`.
        """
        # To avoid division by zero, add a small constant to y_data.
        return y_data + np.finfo(float).eps

    def _calculate_lod_replicate_variance(
        self,
        x_data,
        y_data,
        lower_std_dev_multiplier: float = 2.5,
        upper_std_dev_multiplier: float = 0.0,
    ):
        """
        Calculate the Lower and Upper Limits of Detection (LLOD and ULOD) using variance
        of replicate max and min concentration standards. It ignore zero concentration
        standards. If there are no replicates, the standard deviation zero
        Possible TODO: sometimes a minimum variance is used in other software.

        In the notation below we assume the response signal is the Y-axis and the
        concentration is the X-axis.

        Example: Two replicates of the lowest concentration standard (conc=1.0 pg/ml)
        have standard deviation of 100 across their responses. LLOD will be `model.predict
        (1.0) + 100 * 2.5` where 2.5 is the `lower_std_dev_multiplier` parameter.

        :param bottom_std_dev: Standard deviation at the bottom calibration point.
        :param top_std_dev: Standard deviation at the top calibration point.
        :param std_dev_multiplier: Multiplier for the standard deviations (default 2.5).
        :return: Pair of tuples containing the LLOD and ULOD, and the corresponding x-values.
        """

        x_indexed_y_data = pd.DataFrame({"x": x_data, "y": y_data}).set_index("x")
        # remove zeros from x_data
        x_indexed_y_data = x_indexed_y_data[x_indexed_y_data.index > 0]
        x_min = np.min(x_indexed_y_data.index)
        x_max = np.max(x_indexed_y_data.index)
        bottom_std_dev = x_indexed_y_data.loc[x_min, "y"].std()
        top_std_dev = x_indexed_y_data.loc[x_max, "y"].std()

        # Calculate LLOD and ULOD of RESPONSE SIGNAL
        llod = self.predict(x_min) + (lower_std_dev_multiplier * bottom_std_dev)
        ulod = self.predict(x_max) - (upper_std_dev_multiplier * top_std_dev)

        # Calculate the limits of detection for CONCENTRATION
        llod_x = self.predict_inverse(llod)
        ulod_x = self.predict_inverse(ulod)
        return llod_x, ulod_x, llod, ulod

    def fit(self, x_data, y_data, weight_func=None, LOD_func=None, **kwargs):
        """
        Fit the 4 Parameter Logistic (4PL) model.

        x_data: x data points
        y_data: y data points
        weight_func: optional Function that calculates weights from y_data. This is
            passed into the `curve_fit` function where the function minimized is `sum
            ((r / weight_func(y_data)) ** 2)` where r is the residuals.
            Thus for a typical 1/y^2 weighting, `weight_func` should be `lambda
            y_data: y_data`
        """
        x_data = np.float64(x_data)
        y_data = np.float64(y_data)
        df_data = pd.DataFrame({"x": x_data, "y": y_data})
        df_data.sort_values(by="x", inplace=True)

        if LOD_func is None:
            # default LOD_func is to use replicate variance
            LOD_func = self._calculate_lod_replicate_variance

        absolute_sigma = False
        weights = None
        if weight_func is not None:
            weights = weight_func(y_data)
            absolute_sigma = True

        # Initial guess for the parameters
        self.guess_A_ = np.min(y_data)  # type: ignore
        if self.slope_direction_positive is not None:
            self.guess_B_ = 1.0 if self.slope_direction_positive else -1.0
        else:
            # type: ignore
            self.guess_B_ = (
                1.0
                if np.mean(
                    df_data.iloc[: np.minimum(self.slope_guess_num_points_to_use, len(df_data))][  # type: ignore
                        "y"
                    ]
                )
                < np.mean(
                    df_data.iloc[-np.minimum(self.slope_guess_num_points_to_use, len(df_data)) :][  # type: ignore
                        "y"
                    ]
                )
                else -1.0
            )
        self.guess_C_ = np.mean(x_data)  # type: ignore
        self.guess_D_ = np.max(y_data)  # type: ignore
        initial_guess = [self.guess_A_, self.guess_B_, self.guess_C_, self.guess_D_]

        curve_fit_kwargs = {
            "f": self.four_param_logistic,
            "xdata": x_data,
            "ydata": y_data,
            "p0": initial_guess,
            "maxfev": 10000,
            "sigma": weights,
            "absolute_sigma": absolute_sigma,
        }

        # overwrite parameters with any kwargs passed in
        for k, v in kwargs.items():
            curve_fit_kwargs[k] = v

        # Perform the curve fit
        params, cov = curve_fit(**curve_fit_kwargs)
        self.A_, self.B_, self.C_, self.D_ = params
        self.cov_ = cov
        self.LLOD_, self.ULOD_, self.LLOD_y_, self.ULOD_y_ = LOD_func(x_data, y_data)
        return self

    @staticmethod
    def jacobian(x_data, A, B, C, D):
        """
        Jacobian matrix of the 4PL function with respect to A, B, C, D.
        """
        z = (x_data / C) ** B

        partial_A = 1.0 / (1.0 + z)
        partial_B = -(z * (A - D) * np.log(np.maximum(x_data / C, np.finfo(float).eps))) / (  # type: ignore
            (1.0 + z) ** 2
        )
        partial_C = (B * z * (A - D)) / (C * (1.0 + z) ** 2)
        partial_D = 1.0 - 1.0 / (1.0 + z)

        # Jacobian matrix
        J = np.array([partial_A, partial_B, partial_C, partial_D]).T
        return J

    def predict_confidence_band(self, x_data):
        """
        Predict confidence bands of data points.
        TODO: still need to double-check the math here.

        See:
            https://www.graphpad.com/guides/prism/latest/curve-fitting/reg_graphing_confidence_and_predic.htm
            https://www.graphpad.com/guides/prism/latest/curve-fitting/reg_how_confidence_and_prediction_.htm
            https://stats.stackexchange.com/questions/15423/how-to-compute-prediction-bands-for-non-linear-regression

        """
        if self.cov_ is None:
            raise Exception(
                "Covariance matrix is not available. Please call 'fit' with appropriate data."
            )
        J = self.jacobian(x_data, self.A_, self.B_, self.C_, self.D_)
        pred_var = np.sum((J @ self.cov_) * J, axis=1)

        return np.sqrt(pred_var)

    def predict_prediction_band(self, x_data, y_data):
        """
        Predict prediction bands of data points.
        TODO: still need to double-check the math here.
        """
        ss = (y_data - self.predict(x_data)) ** 2
        df = len(x_data) - 4  # 4 parameters

        return np.sqrt(self.predict_confidence_band(x_data) ** 2 * ss / df)

    def predict_inverse(self, y):
        # TODO: could merge this with `predict` and make it a parameter
        """Inverse 4 Parameter Logistic (4PL) model.

        Used for calculating the x-value for a given y-value.
        Usually, standard curves are fitted using concentration as x-values and response as
        y-values, so that variance in response is modeled for a given known concentration.
        But for samples of unknown concentration, we want to get the concentration as given
        response, which is what this function does.

        """
        self.check_fit()
        z = ((self.A_ - self.D_) / (y - self.D_)) - 1  # type: ignore

        # For addressing fractional powers of negative numbers, np.sign(z) * np.abs(z) used rather than z
        # https://stackoverflow.com/questions/45384602/numpy-runtimewarning-invalid-value-encountered-in-power
        return self.C_ * (np.sign(z) * np.abs(z) ** (1 / self.B_))  # type: ignore

    def predict(self, x_data):
        self.check_fit()
        return self.four_param_logistic(x_data, self.A_, self.B_, self.C_, self.D_)
